### PR TITLE
integrate private dns

### PR DESCRIPTION
**Pull request by alexnl022**
From @jpuchois Tea, solving issue https://github.com/octodns/octodns-azure/issues/47.
I'm not a Python expert, I've just make the upgrade, test it on our infrastructure, & add test to validate requirements. https://github.com/octodns/.github/blob/7aadbc5563c635115646aeddb9e6c20c4a7f9dce/CONTRIBUTING.md
Regards,

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -7,11 +7,12 @@ azure-core==1.26.3
 azure-identity==1.7.1
 azure-mgmt-core==1.3.2
 azure-mgmt-dns==8.0.0
+azure-mgmt-privatedns==1.0.0
 azure-mgmt-trafficmanager==1.0.0
 certifi==2022.12.7
 cffi==1.15.1
 charset-normalizer==3.0.1
-cryptography==39.0.0
+cryptography==39.0.1
```
Comment (ross) [inline requirements.txt:15]: I assume these updates were pulled in by `./script/update-requirements` and not anything strictly required?

**Review by ross — COMMENTED** [review]
Annoying how similar the two APIs are, but slightly different. 

My only high level question is whether it'd make sense to have something like `AzureBaseProvider`  with `AzurePrivateProvider` and `AzureProvider` inheriting from it rather than a param to drive the difference. 

I think that'd essentially Look like renaming `AzureProvider` to `AzureBaseProvider` and then replacing all the `if not self.private_dns:` bits here with calls to methods that were implemented in the two child classes.  It's a little bit of a preference thing, but it seems like the 2 classes would be more explicit/clear in configs and result in a little bit clearer/more readable code. Thoughts?

File: requirements.txt (hunk 1 above)
Comment (alexnl022) [reply requirements.txt:15]: exatcly :),  I think I followed all the instructions :)

**Comment by alexnl022** [comment]
@ross I think it would be great, also with a split of the code by type, and reshape "utils" functions by type (string aggregator, health-check, validator ....). On our side we needed to have this feature so we worked on it, and I also spent some time to prepare the MR (Fonctional Test, Unitest...). Having more time on our side seems a bit difficult.
Seeing that we have very few lines changed in __init__.py (only 120), would it be possible to merge it ? I think other person would be glad to have this feature integrate.

**Comment by jpuchois** [comment]
Hello @ross ,

Any news on this ? We know it's kind of perfectible but is it possible to merge like this ?

Thanks a lot

**Comment by viranch** [comment]
I'm yet to look at the code change but I agree with @ross in theory. I'll take a deeper dive asap.

**Comment by ross** [comment]
> We know it's kind of perfectible but is it possible to merge like this ?

I'd rather not merge something and then substantially change the way it works afterwards. Anyone that needs to use it should be able to do so via a pinned SHA, e.g. https://github.com/octodns/octodns-azure#shas.

Assuming you all don't have the time to make the refactors I can make a pass through to do it. I will only be able to verify the non-private setup though so someone else will have to check that the new stuff works as desired.

**Comment by viranch** [comment]
> Assuming you all don't have the time to make the refactors I can make a pass through to do it. I will only be able to verify the non-private setup though so someone else will have to check that the new stuff works as desired.

I can spend some time on it this week including testing stuff.

**Comment by ross** [comment]
Quick pass at the refactoring over in https://github.com/octodns/octodns-azure/pull/50 ready for testing/:eyes:

**Pull request merged**